core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+
+      if (this.formatVersion == newFormatVersion) {
+        return this;
+      }
+
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      Preconditions.checkArgument(null != newLocation, "Invalid location: null");
+      if (null != this.location && this.location.equals(newLocation)) {
+        return this;
+      }
+
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      this.currentVersionId = newVersionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder setSchemas(Iterable<Schema> schemasToAdd) {
+      int highestFieldId =
+          Lists.newArrayList(schemasToAdd).stream()
+              .map(Schema::highestFieldId)
+              .max(Integer::compareTo)
+              .orElse(0);
+      for (Schema schema : schemasToAdd) {
+        this.schemas.add(schema);
+        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
+      }
+
+      return this;
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
+      this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(version.timestampMillis())
+              .versionId(version.versionId())
+              .build());
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> updated) {
+      if (updated.isEmpty()) {
+        return this;
+      }
+
+      this.properties.putAll(updated);
+      this.changes.add(new MetadataUpdate.SetProperties(updated));
+      return this;
+    }
+
+    public Builder removeProperties(Set<String> propertiesToRemove) {
+      if (propertiesToRemove.isEmpty()) {
+        return this;
+      }
+
+      propertiesToRemove.forEach(this.properties::remove);
+      this.changes.add(new MetadataUpdate.RemoveProperties(propertiesToRemove));
+      return this;
+    }
+
+    private static Map<Integer, ViewVersion> indexVersions(List<ViewVersion> versionsToIndex) {
+      ImmutableMap.Builder<Integer, ViewVersion> builder = ImmutableMap.builder();
+      for (ViewVersion version : versionsToIndex) {
+        builder.put(version.versionId(), version);
+      }
+
+      return builder.build();
+    }
+
+    private static Map<Integer, Schema> indexSchemas(List<Schema> schemasToIndex) {
```

Comment:
Why are these static methods inside the builder?

Reply:
we're using them in `ViewMetadata` + in the `Builder` and making them static is the only way to share the logic for both

Reply:
Sure, but why are they inside the builder? What is the purpose of scoping these methods to the builder and then using them outside?

Reply:
Ah, I see. Because this is using Immutables, the outer scope is an interface where helper methods can't be made private.